tests/frontend/compilers/test_compiler.py

```diff
@@ -0,0 +1,130 @@
+        """If a circuit spec includes a Blackbird program,
+        the topology property should return the equivalent
+        directed acyclic graph.
+        """
+
+        class DummyCircuit(Compiler):
+            """Dummy circuit used to instantiate
+            the abstract base class"""
+            modes = 0
+            remote = False
+            local = True
+            interactive = True
+            primitives = set()
+            decompositions = set()
+
+            circuit = textwrap.dedent(
+                """\
+                name test
+                version 0.0
+
+                Sgate(0.543, 0) | 0
+                Dgate(-7.123) | 1
+                BSgate(0.54) | 0, 1
+                MeasureFock() | 0
+                MeasureFock() | 2
+                """
+            )
+
+        dummy = DummyCircuit()
+        top = dummy.graph
+
+        circuit = top.nodes().data()
+
+        # check circuit is correct length
+        assert len(circuit) == 5
+
+        # check gates are correct
+        assert circuit[0]['name'] == 'Sgate'
+        assert circuit[1]['name'] == 'Dgate'
+        assert circuit[2]['name'] == 'BSgate'
+        assert circuit[3]['name'] == 'MeasureFock'
+        assert circuit[4]['name'] == 'MeasureFock'
+
+        # check topology/edges between nodes
+        edges = {(i, j) for i, j, d in top.edges().data()}
+        assert edges == {(0, 2), (1, 2), (2, 3)}
+
+    def test_template_topology_construction(self):
+        """If a circuit spec includes a Blackbird template,
+        the topology property should return the equivalent
+        directed acyclic graph, with all parameters set to zero.
+        """
+
+        class DummyCircuit(Compiler):
+            """Dummy circuit used to instantiate
+            the abstract base class"""
+            modes = 0
+            remote = False
+            local = True
+            interactive = True
+            primitives = set()
+            decompositions = set()
+
+            circuit = textwrap.dedent(
+                """\
+                name test
+                version 0.0
+
+                Sgate({sq}, 0) | 0
+                Dgate(-7.123) | 1
+                BSgate({theta}) | 0, 1
+                MeasureFock() | 0
+                MeasureFock() | 2
+                """
+            )
+
+        dummy = DummyCircuit()
+        top = dummy.graph
+
+        circuit = top.nodes().data()
+
+        # check circuit is correct length
+        assert len(circuit) == 5
+
+        # check gates are correct
+        assert circuit[0]['name'] == 'Sgate'
+        assert circuit[1]['name'] == 'Dgate'
+        assert circuit[2]['name'] == 'BSgate'
+        assert circuit[3]['name'] == 'MeasureFock'
+        assert circuit[4]['name'] == 'MeasureFock'
+
+        # check arguments
+        assert circuit[0]['args'] == [0, 0]
+        assert circuit[1]['args'] == [-7.123]
+        assert circuit[2]['args'] == [0]
```

Comment:
might be missing a bit here: how did the `{sq}` and `{theta}` symbolic parameters get numeric values?

Reply:
The line `top = dummy.graph` excecutes:
```python
if bb.is_template():
            params = bb.parameters
            kwargs = {p: 0 for p in params}
```
inside of `compiler.graph` which seems to set all parameters to `0`.

Reply:
could it be a side effect there? :thinking: